### ISSUE TITLE
Pass in width and height for graphs to use for info density tuning

### ISSUE THREAD

**Issue by brendanheywood** [theCrag/website#2615]
The width of the graph in the profile page context is smaller than what I was testing with. I'd also like to put the graph onto the ascents page when their is only a single person selected. In this context the graph will be almost full width. If the graph was dynamically generated it could be responsive and figure out how wide things should be but even then the template data would still be the same

So what I would like is for the graph url to be passed in a height and width which is then:

* [ ] used by the backed to figure out how many time slices there should be. I have redone the algorithm which selects which time granularity to use so it simply starts with a desired number of slices and work from there. The number of slices should be directly related to the pixel width, so in a large embed we get more bars than a smaller embed so roughly they have the same pixel width for each slice. See my comments in Tools.pm line in selectDefaultTimeSeries()

**Comment by brendanheywood** [comment]
If someone has just joined up and logged their first tick, we do not want to show a graph which only has a single day on the time axis and a single grade on the y axis. This still applies even after a few ticks where there may be a tight cluster in both time and difficulty.

Currently we show this - should a limit of 10 still apply? btw I just tried to create some test data but the dev db is bork missing some new fields so things explode. Someone with 9 ticks will still get cpr, and still get ranked, so I'm not sure about the graphs. The nice thing about the 10 minimum is that it might encourage people to tick more, but at the same time it doesn't really help show them what the plaftorm is capable of. So I'm leaning towards turning on the graphs after the first tick.

btw here is some test cases which have the bare minimum of 10 ticks to trigger the graphs:

https://dev.thecrag.com/ascents/aggregate/graph-sport-cpr/with-route-gear-style/sport/by/ameliaclimbs/

https://dev.thecrag.com/ascents/aggregate/graph-sport-cpr/with-route-gear-style/sport/by/oshuaj33/

> You have not logged enough ascents to show all the profile features. Log 10 ascents to unlock them.

So what I think we should do is:

* [ ] when selecting a time series, choose the best granularity, but then if we are way short on the desired number of sliced make some empty slices going back in time. For instance right now in my revised algorithm for choosing slices I've hard coded 28 as the minimum # of slices, but if you end up selecting days then show a graph with 28 days of empty data and only the last few days filled in. I think this special padding logic would only apply to days, and maybe weeks but nothing larger.

* [ ] the same should apply to the grade scale. If we have only ticks say 18-20, the system should always show a minimum grade range of say 200 internal grades. It should try to show the grades you ticked in the middle of this band, and if either the lower or upper range exceeds 0 or 500 then move the whole band up or down accordingly (maybe you do this limiting already?)

**Comment by scd** [comment]
I agree with both of these suggestions and will make these changes.

 - [ ] Also if there are no ascents we should do an empty graph - Simon to provide default x and y axis.

**Comment by scd** [comment]
I have added a height and width to the url parameters which filter through to both the period selection algorithm and the template. Seems to work great.

@brendanheywood the period selection min of 28 has been changed to int(width/18) which gives 28 for the default width of 28. Do we need to make this a bit smarter (ie take into account the fixed width of the grades)?

I have also put in place a minimum time slices and a minimum grades. I have not implemented exactly as you suggested, but there is enough there to start to talk about.

Closing :)
